### PR TITLE
Implement 'aea get' and 'aea set' to read/write from/to AEA configuration file

### LINKED ISSUE #1670: Add running agent configuration in a separate YAML object
**Is your feature request related to a problem? Please describe.**
In `aea-config.yaml`, there are many fields related to parameters for the runtime performances of an AEA (e.g. runtime_mode, loop_mode etc.).

**Describe the solution you'd like**
Would be useful to have a specific object that contains all these parameters. I think it makes the AEA configuration more readable.

**Describe alternatives you've considered**
n/a

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

Implement 'aea get' and 'aea set' to read/write from/to AEA configuration file.

This PR also fixes what fields are considered `config` versus `description`. Fields are `config` if changing them does not necessarily imply a change in code.

## Fixes

#1670

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

n/a

### CODE REVIEW

File: aea/cli/config.py [hunk 1]
```diff
@@ -39,11 +48,11 @@ def config(click_context):  # pylint: disable=unused-argument
 
 
 @config.command()
-@click.argument("JSON_PATH", required=True, type=AEAJsonPathType())
+@click.argument("JSON_PATH", required=True)  # , type=AEAJsonPathType())
```
Comment (DavidMinarsch) [inline aea/cli/config.py:51]: why commented out? remove if no longer needed

File: aea/cli/config.py [hunk 2]
```diff
@@ -54,89 +63,269 @@ def get(ctx: Context, json_path: List[str]):
...
+        "type",
+        "connections",
+        "skills",
+        "protocols",
+        "contracts",
+    ]
 
-    configuration_object = load_yaml(configuration_file_path)
-    config_loader.validator.validate(instance=configuration_object)
+    AGENT_FIELDS_ALLOWED_TO_CHANGE = [
```
Comment (DavidMinarsch) [inline aea/cli/config.py:93]: These should be defined on the relevant `ComponentConfiguration/PackageConfiguration`/their derived classes.

Comment (DavidMinarsch) [reply aea/cli/config.py:93]: I think you should do whitelisting instead. that's much simpler to maintain

File: aea/cli/config.py [hunk 3]
```diff
@@ -54,89 +63,269 @@ def get(ctx: Context, json_path: List[str]):
...
+                and component_id.package_type == self.component_id.package_type
+                and component_id.name == self.component_id.name
+            ):
+                break
+        else:  # pragma: nocover  # should be always ok, cause component has to be alrady registered
+            raise ValueError("component is not registered?")
+
+        self.component_id = component_id
+
+    def _parent_object_for_agent_component_configuration(self) -> Dict:
```
Comment (DavidMinarsch) [inline aea/cli/config.py:252]: Missing docstring, incomplete typing

File: aea/cli/config.py [hunk 4]
```diff
@@ -54,89 +63,269 @@ def get(ctx: Context, json_path: List[str]):
...
+            )
+            self._update_object(parent_configuration_object, type_str, value)
+            agent_configuration_object = self.agent_config.json
+        else:
+            # already agent
+            self._update_object(parent_configuration_object, type_str, value)
+            agent_configuration_object = configuration_object
+        self._dump_agent_configuration(agent_configuration_object)
+
+    def _update_object(self, parent_object: Dict, type_str: str, value: str) -> None:
```
Comment (DavidMinarsch) [inline aea/cli/config.py:298]: incomplete docstring

File: aea/cli/config.py [hunk 5]
```diff
@@ -54,89 +63,269 @@ def get(ctx: Context, json_path: List[str]):
...
+            self.is_target_agent
+            and self.json_path[0] in self.AGENT_FIELDS_ALLOWED_TO_CHANGE
+        ):
+            return
+        if self.json_path[0] in self.FIELDS_NOT_ALLOWED_TO_CHANGE:
+            raise click.ClickException(
+                f"Field `{self.json_path[0]}` is not allowed to change!"
+            )
+
+    def _fix_component_id_version(self) -> None:
```
Comment (DavidMinarsch) [inline aea/cli/config.py:233]: why is this needed?

File: aea/cli/config.py [hunk 6]
```diff
@@ -54,89 +63,269 @@ def get(ctx: Context, json_path: List[str]):
...
-        raise click.ClickException("Attribute '{}' not found.".format(attr_name))
-    if not isinstance(parent_obj.get(attr_name), (str, int, bool, float)):
-        raise click.ClickException(
-            "Attribute '{}' is not of primitive type.".format(attr_name)
+
+    def _get_value_from_configuration_object(self, conf_obj: Dict) -> Any:
+        """Get value from configuration object."""
+        return self._get_parent_object(conf_obj).get(self.attr_name)
+
+    def _get_parent_object(self, conf_obj: Dict) -> Dict:
```
Comment (DavidMinarsch) [inline aea/cli/config.py:186]: dict needs to be fully typed

File: aea/cli/config.py [hunk 7]
```diff
@@ -54,89 +63,269 @@ def get(ctx: Context, json_path: List[str]):
...
-        raise click.ClickException("Cannot convert {} to type {}".format(value, type_))
+            configuration_object = load_yaml(str(self.path_to_resource_configuration))
 
-    try:
-        configuration_obj = config_loader.configuration_class.from_json(
-            configuration_object
+        self.config_loader.validate(configuration_object)
+        return configuration_object
+
+    def _get_configuration_object_from_agent(self) -> Optional[Dict]:
```
Comment (DavidMinarsch) [inline aea/cli/config.py:173]: dict needs to be fully typed

File: aea/cli/config.py [hunk 8]
```diff
@@ -54,89 +63,269 @@ def get(ctx: Context, json_path: List[str]):
...
+    @property
+    def is_target_agent(self) -> bool:
+        """Is target of get/update is agent config."""
+        return self.component_id is None
 
-    type_ = FROM_STRING_TO_TYPE[type_str]
-    try:
-        if type_ != bool:
-            parent_object[attribute_name] = type_(value)
+    def _load_configuration_object(self) -> Dict:
```
Comment (DavidMinarsch) [inline aea/cli/config.py:163]: dict needs to be fully typed

File: aea/cli/config.py [hunk 9]
```diff
@@ -54,89 +63,269 @@ def get(ctx: Context, json_path: List[str]):
...
+            return handle_dotted_path(self.dotted_path, self.agent_config.author)
+        except AEAException as e:
+            raise click.ClickException(*e.args)
+
+    @property
+    def attr_name(self) -> str:
+        """Attribute name."""
+        return self.json_path[-1]
+
+    def get(self) -> Any:
```
Comment (DavidMinarsch) [inline aea/cli/config.py:133]: rather than Any, can you define the union of valid types please?

File: aea/cli/config.py (hunk 1 above)
Comment (solarw) [reply aea/cli/config.py:51]: not required anymore. removed

File: aea/cli/config.py (hunk 5 above)
Comment (solarw) [reply aea/cli/config.py:233]: cause component id created with latest version, not current version

File: aea/cli/config.py (hunk 6 above)
Comment (solarw) [reply aea/cli/config.py:186]: i can type it cause it component configuration and it always differs

File: aea/cli/config.py (hunk 3 above)
Comment (solarw) [reply aea/cli/config.py:252]: i can not type dict object, cause it depends  configuration and varibale

File: aea/cli/config.py (hunk 7 above)
Comment (solarw) [reply aea/cli/config.py:173]: type can not be specified, configurations differs

File: aea/cli/config.py (hunk 8 above)
Comment (solarw) [reply aea/cli/config.py:163]: configuration differs

File: aea/cli/config.py (hunk 4 above)
Comment (solarw) [reply aea/cli/config.py:298]: done

File: aea/cli/config.py (hunk 9 above)
Comment (solarw) [reply aea/cli/config.py:133]: it uses dict get operation, there cane be any value, depends on how the component config was loaded. it's ok to assume you dont know the type is going to be returned

File: aea/cli/config.py (hunk 2 above)
Comment (solarw) [reply aea/cli/config.py:93]: actually this check probably is redundant, cause there is a schema check on save.

File: aea/cli/config.py [hunk 10]
```diff
@@ -54,89 +63,269 @@ def get(ctx: Context, json_path: List[str]):
...
+    ConfigGetSet(ctx, json_path).set(value, type)
+
 
+class ConfigGetSet:
+    """Tool to get/set value in agent config."""
 
-def _get_config_value(ctx: Context, json_path: List[str]):
-    config_loader = cast(ConfigLoader, ctx.config.get("configuration_loader"))
-    configuration_file_path = cast(str, ctx.config.get("configuration_file_path"))
+    FIELDS_NOT_ALLOWED_TO_CHANGE = [
```
Comment (DavidMinarsch) [inline aea/cli/config.py:82]: See previous comments; should be moved to config!

File: aea/configurations/base.py [hunk 11]
```diff
@@ -1867,6 +1889,10 @@ class ContractConfig(ComponentConfiguration):
     default_configuration_filename = DEFAULT_CONTRACT_CONFIG_FILE
     package_type = PackageType.CONTRACT
 
+    FIELDS_ALLOWED_TO_UPDATE = [
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:1892]: Actually no. this should not be configurable as the interface is hard coded in the package.

File: aea/configurations/base.py [hunk 12]
```diff
@@ -1370,6 +1382,16 @@ class AgentConfig(PackageConfiguration):
...
+        "agent_name",
+        "author",
+        "registry_path",
+        "logging_config",
+        "private_key_paths",
+        "default_connection",
+        "default_ledger",
+    ]
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:1393]:
```suggestion
    FIELDS_ALLOWED_TO_UPDATE = [
        "registry_path",
        "logging_config",
        "private_key_paths",
        "connection_private_key_paths",
    ]
```

File: aea/configurations/base.py [hunk 13]
```diff
@@ -1178,6 +1184,12 @@ class SkillConfig(ComponentConfiguration):
     default_configuration_filename = DEFAULT_SKILL_CONFIG_FILE
     package_type = PackageType.SKILL
 
+    FIELDS_ALLOWED_TO_UPDATE = [
+        "behaviours",
+        "handlers",
+        "models",
+    ]
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:1191]: actually; only the "args" field nested under these fields is configurable...

File: aea/cli/config.py (hunk 10 above)
Comment (solarw) [reply aea/cli/config.py:82]: moved

File: aea/configurations/base.py (hunk 13 above)
Comment (solarw) [reply aea/configurations/base.py:1191]: it's done with schema validation, and actually  works

File: aea/configurations/base.py (hunk 11 above)
Comment (solarw) [reply aea/configurations/base.py:1892]: need to discuss,

going to stay as is, but as i told above, schema check does the trick, so this is presave check and provide some flexibility i'm not sure we need

File: aea/configurations/base.py (hunk 12 above)
Comment (solarw) [reply aea/configurations/base.py:1393]: done

File: aea/configurations/base.py (hunk 11 above)
Comment (solarw) [reply aea/configurations/base.py:1892]: schema does check for new fields, but allows to modify already defined, so it does not cover all the cases. so this check still required

File: aea/cli/config.py (hunk 10 above)
Comment (DavidMinarsch) [reply aea/cli/config.py:82]: Done

**Comment by marcofavorito** [comment]
Set up:
- run the following commands:
```
aea create myagent && cd myagent
aea add skill fetchai/echo
aea scaffold skill my_skill
```
- append the following text to the configuration file:
```
---
name: echo
author: fetchai
version: 0.8.0
type: skill
behaviours:
  echo:
    args:
      tick_interval: 2.0
---
name: my_skill
author: default_author
version: 0.1.0
type: skill
behaviours:
  scaffold:
    args:
      foo: bar2  # original value was 'bar'
```


Get commands:
- [ ] get field from main body of `aea-config.yaml` (descriptive or configuration)
  `aea config get agent.agent_name`
- [ ] get field from component configuration of `aea-config.yaml` (descriptive or configuration)  
- [ ] get field from vendor component: `aea config get vendor.fetchai.skills.echo.behaviours.echo.args.tick_interval
` gives `2.0`
- [ ] get field from custom component: `aea config get skills.my_skill.behaviours.scaffold.args.foo` gives `bar2`

Set commands:
TODO
